Fix calculation of new block index on drop

oandregal opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/9823, https://github.com/WordPress/gutenberg/issues/10202 and https://github.com/WordPress/gutenberg/issues/9303.

### Testing

- Drag any block and drop it at the place you want. Check that it behaves as expected.
- Drag any block and drop it ot its own top/bottom zones ([see](https://github.com/WordPress/gutenberg/issues/10202)).
- Drag a columns block and drop it within one of its columns ([see](https://github.com/WordPress/gutenberg/issues/9303#issuecomment-420563076)).
- Drag a block and drop it in the top zone of any block below it ([see](https://github.com/WordPress/gutenberg/issues/9823)).

oandregal committed:
Remove unnecessary layout prop

oandregal committed:
Fix bug on dropping top-level blocks

oandregal committed:
Improve readibility

nosolosw commented on packages/editor/src/components/block-draggable/index.js:
```diff
@@ -4,13 +4,12 @@
-		rootClientId,
-		clientId,
-		layout,
+		srcIndex: index,
```
Improve nomenclature, as it was inconsistent (`fromIndex` but not `fromClientId`?), especially when this data was used in the `BlockDropZone`.

oandregal committed:
Do not move block if it drops on itself

oandregal committed:
Extract getDescendants selector

oandregal committed:
Prevent blocks from moving:

- if it is drop into one of its dropzones
- if it is drop into one of its descendant blocks' dropzones

oandregal committed:
Make code more readable

youknowriad reviewed (commented):
Tested and works properly, I'd appreciate if we can just change the name of `getDescendants`.

Also, it's important to run `npm run docs:build` to regenerate the new selector's docs (and commit the changes).

oandregal committed:
Rename getDescendants to getClientIdsOfDescendants

nosolosw commented on packages/editor/src/store/selectors.js:
```diff
@@ -565,6 +565,20 @@ export const getBlocks = createSelector(
+ *
+ * @return {Array} ids of descendants.
+ */
+export const getDescendants = ( state, clientIds ) => flatMap( clientIds, ( clientId ) => {
```
Renamed it to `getClientIdsOfDescendants` in bebfdbd.

youknowriad reviewed (approved):
LGTM 👍

nosolosw commented on packages/editor/src/components/block-drop-zone/index.js:
```diff
@@ -56,28 +77,29 @@ class BlockDropZone extends Component {
+			// so we also need to account for that case separately.
+			return ( srcRoot === dstRoot ) || ( ! srcRoot === true && ! dstRoot === true );
+		};
+		const isSameBlock = ( src, dst ) => src === dst;
```
I know it seems straightforward code, but I like how having the function helps document the intent for the reader.
